### ISSUE TITLE
the usage of `^.` seems redundant

### ISSUE THREAD

**Issue by yegor256** [objectionary/eo#3481]
We use `^.` too often in `.eo` files. In the majority of situations this prefix may be skipped and the location of the object will be automatically found by the compiler.

**Comment by yegor256** [comment]
@maxonfjvipon compare this:

```
[] > foo
  [k] > bar
  bar 42 > x
```

and this:

```
[] > foo
  [k] > bar
  ^.bar 42 > x
```

The functionality here is not the same. Am I right?

**Comment by maxonfjvipon** [comment]
@yegor256 yes, they are different

**Comment by yegor256** [comment]
@maxonfjvipon I guess, the first snippet must be written as:

```
[] > foo
  [k] > bar
  $.bar 42 > x
```

Thus, we either have `\xi.bar` (`$.bar`) or `\xi.\rho.bar` (`^.bar`). Just `bar` means the first option. Am I right?

**Comment by maxonfjvipon** [comment]
@yegor256 yes, the first one

**Comment by yegor256** [comment]
@maxonfjvipon maybe it would be better to make our syntax more explicit: `bar` means `$.bar` and that's it. If `^.bar` is meant, it should be written as such. No defaults. WDYT?

**Comment by maxonfjvipon** [comment]
@yegor256 I'm not sure because it'll make the language less convenient for end user (and I think we want it to be convenient and we actually add such syntax sugar, for example `>>`), but it opposite it'll make it more strict

**Comment by maxonfjvipon** [comment]
@yegor256 Actually yes, I think it's a good idea

**Comment by yegor256** [comment]
@maxonfjvipon maybe the other way around would be better:

`$.bar` means `\xi.bar` (we use it rarely, as I see)

`bar` means `\xi.\rho.\rho...bar` (where it is found)

**Comment by maxonfjvipon** [comment]
@yegor256 it should already work like this

**Comment by yegor256** [comment]
@maxonfjvipon  not really. If I use just `bar` it means `\xi.bar`:

```
[] > foo
  [k] > bar
  bar 42 > x
```

**Comment by maxonfjvipon** [comment]
@yegor256 yes, if `bar` exists in the same scope, if it's not it'll try to find `^.bar`, `^.^.bar` and so on

**Comment by yegor256** [comment]
@maxonfjvipon this code should not compile:

```
[] > foo
  [k] > bar
  bar 42 > x
```

instead, it should be written as such:

```
[] > foo
  [k] > bar
  $.bar 42 > x
```

**Comment by 0pdd** [comment]
@yegor256 2 puzzles [#3814](https://github.com/objectionary/eo/issues/3814), [#3815](https://github.com/objectionary/eo/issues/3815) are still not solved.

**Comment by 0pdd** [comment]
@yegor256 the puzzle [#3815](https://github.com/objectionary/eo/issues/3815) is still not solved; solved: [#3814](https://github.com/objectionary/eo/issues/3814), [#3869](https://github.com/objectionary/eo/issues/3869).

**Comment by 0pdd** [comment]
@yegor256 all 3 puzzles are solved here: [#3814](https://github.com/objectionary/eo/issues/3814), [#3815](https://github.com/objectionary/eo/issues/3815), [#3869](https://github.com/objectionary/eo/issues/3869).
